Add support of adding routes

cathay4t opened the pull request:
Only support append routes to existing configuration for now.

tyll commented on libnmstate/nm/route.py:
```diff
@@ -141,3 +149,40 @@ def _get_default_route_config(gateway, metric, default_table_id, iface_name):
+        else socket.AF_INET
+    metric = route.get(Route.METRIC, NM_ROUTE_DEFAULT_METRIC)
+    next_hop = route[Route.NEXT_HOP_ADDRESS]
+    rt = nmclient.NM.IPRoute.new(family, destination, prefix_len, next_hop,
```
IMHO it is better to not abbreviate the variable here and use `route`.

cathay4t replied:
Changed to `ip_route` for `NM.IPRoute`.

tyll commented on libnmstate/state.py:
```diff
@@ -159,6 +217,42 @@ def merge_interfaces(self, other_state):
+    def merge_route_config(self, other):
+        """
+        Merge routes from other to self.
+        If interface does exists in self, create an empty one.
```
I do not understand the sentence above and it is also not correct English (`does exist` without `s`) and there needs to be an article (an or the) before `interface`.

cathay4t replied:
Changed to:
```txt
If any changed route referring non-exist interface, create the
interface state with name only, in order to trigger the recreating of
the interface profile.
```

cathay4t commented:
Amended patch uploaded with variable name and comment changes.

tyll commented on libnmstate/metadata.py:
```diff
@@ -54,13 +57,16 @@ def generate_ifaces_metadata(desired_state, current_state):
         iface_state.pop('_master', None)
         iface_state.pop('_master_type', None)
         iface_state.pop('_brport_options', None)
+        iface_state.get(Interface.IPV4, {}).pop(Route.KEY, None)
```
The other metadata is in a namespace with underscores... Maybe this should be considered here as well. On the other hand using these metadata fields seems to me to be just a workaround instead of making it a real objects.

tyll commented on libnmstate/netapplier.py:
```diff
@@ -104,10 +103,11 @@ def _choose_checkpoint(dbuspath):
 def _apply_ifaces_state(desired_state, verify_change, commit,
                         rollback_timeout):
-    current_state = state.State({Constants.INTERFACES: netinfo.interfaces()})
+    current_state = state.State(netinfo.show())
```
This seems to be an unrelated change. It would be better to do them separately in the future.

cathay4t replied:
This is related as `State` now holds the route configuration also.

tyll commented on libnmstate/netapplier.py:
```diff
@@ -116,9 +116,7 @@ def _apply_ifaces_state(desired_state, verify_change, commit,
-                current_state = state.State(
-                    {Constants.INTERFACES: netinfo.interfaces()}
-                )
+                current_state = state.State(netinfo.show())
```
Also seems to be an unrelated change

cathay4t replied:
This is related as State now holds the route configuration also.

tyll commented on libnmstate/netapplier.py:
```diff
@@ -129,8 +127,9 @@ def _apply_ifaces_state(desired_state, verify_change, commit,
 
 def _verify_change(desired_state):
-    current_state = state.State({Constants.INTERFACES: netinfo.interfaces()})
+    current_state = state.State(netinfo.show())
```
Another unrelated change, is it not?

cathay4t replied:
This is related as State now holds the route configuration also.

tyll replied:
This is a good reason. Would be great to mention these kind of reasons in the commit message - it could have been a separate commit as well in preparation for the route changes.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -31,6 +31,10 @@ def create_setting(config, base_con_profile):
             setting_ipv4.props.ignore_auto_routes = False
             setting_ipv4.props.never_default = False
             setting_ipv4.props.ignore_auto_dns = False
+            setting_ipv4.props.gateway = None
```
As a future refactoring step it would be great to make the ip settings objects and have the common code in a base class that ipv4/ipv6 can then inherit from.

cathay4t replied:
Thanks. I will consider it in another PR.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -51,6 +55,8 @@ def create_setting(config, base_con_profile):
                 nmclient.NM.SETTING_IP4_CONFIG_METHOD_MANUAL)
             _add_addresses(setting_ipv4, config['address'])
+
+        nm_route.add_routes(setting_ipv4, config.get('routes', []))
```
This should probably be a constant.

cathay4t replied:
Changed to `_routes` as you suggested for meta data name.

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -116,6 +120,7 @@ def create_setting(config, base_con_profile):
         setting_ip.props.method = (
             nmclient.NM.SETTING_IP6_CONFIG_METHOD_LINK_LOCAL)
 
+    nm_route.add_routes(setting_ip, config.get('routes', []))
```
Should this be a constant? `routes` I mean.

cathay4t replied:
Changed to _routes as you suggested for meta data name.

tyll commented on libnmstate/metadata.py:
```diff
@@ -147,3 +153,22 @@ def _generate_link_master_metadata(ifaces_desired_state,
+def _generate_route_metadata(desired_state):
+    """
+    Save routes under interface for creating setting profile.
+    Currently route['next-hop-interface'] is mandatory.
```
The comments here a not completely clear. After a lot of thinking I understood them but maybe you can rephrase them. It is irritating that the comment mentions the next hop interface but it is not explicitly used in the code AFAIU it is used by the iface_routes property. One guess to make it cleaner would be to have InterfaceState objects that handle the full state and properly generate them from the dict when the actual state object is created.

On second thought, maybe this function could be a method of the State class that is called in the constructor instead.

cathay4t replied:
Comment changed to:

```
Save routes under interface IP protocol so that nm/ipv4.py or nm/ipv6.py
could include route configuration in `create_setting()`.
Currently route['next-hop-interface'] is mandatory.
```

`InterfaceState` is too big for this PR. If possible, we need a new PR for the class/object
represented states.

cathay4t commented:
@tyll New patch uploaded.

Changes:
 * Changed the metadata key name to `_routes`.
 * Improved the route merging function.
 * Added the example for adding routes `examples/eth1_add_route.yml`.

EdDev commented on libnmstate/metadata.py:
```diff
@@ -147,3 +153,25 @@ def _generate_link_master_metadata(ifaces_desired_state,
+            continue
+        iface_state = desired_state.interfaces.get(iface_name, {})
+        for family in (Interface.IPV4, Interface.IPV6):
+            if family in iface_state:
```
Consider assuming these keys exists already: https://github.com/nmstate/nmstate/pull/340

cathay4t replied:
Will do after #340 got merged.

tyll replied:
#340 is merged

EdDev commented on libnmstate/metadata.py:
```diff
@@ -147,3 +153,25 @@ def _generate_link_master_metadata(ifaces_desired_state,
+    for iface_name, routes in six.viewitems(desired_state.iface_routes):
+        if not iface_name:
+            continue
+        iface_state = desired_state.interfaces.get(iface_name, {})
```
In the "merge" step the desired state has been already complemented with the interfaces of new routes.
Can we drop the question about the existence of the interface in the state?

cathay4t replied:
Done.

tyll replied:
I think @EdDev meant to not use `.get()` but `[iface_name]`.

cathay4t commented:
> I'm missing tests for these changes: nm and libnmstate integration tests.

The test will be included in another PR #334 as it require cleanup(remove routes). And I don't want wast time on creating test cases for this mid-stage PR.

cathay4t commented on libnmstate/metadata.py:
```diff
@@ -147,3 +153,25 @@ def _generate_link_master_metadata(ifaces_desired_state,
+    Currently route['next-hop-interface'] is mandatory.
+    """
+    for iface_name, routes in six.viewitems(desired_state.iface_routes):
+        if not iface_name:
```
We support `state: absent` without interface defined.

cathay4t commented on libnmstate/metadata.py:
```diff
@@ -54,13 +57,16 @@ def generate_ifaces_metadata(desired_state, current_state):
         iface_state.pop('_master', None)
         iface_state.pop('_master_type', None)
         iface_state.pop('_brport_options', None)
+        iface_state.get(Interface.IPV4, {}).pop('_routes', None)
```
Will do once #341 got merged.

cathay4t commented:
New patch uploaded with fix for https://github.com/nmstate/nmstate/pull/331#discussion_r281029956

cathay4t commented:
New patch uploaded.

Changes:
 * Integration test included at `tests/integration/route_test.py` for adding route test.
 * Use constants for route metadata.
 * Previous patch has confirmed interface state always has 'ipv4' and 'ipv6' entry,
   hence changed `iface_state.get()` to `iface_state[]`.

EdDev commented on libnmstate/state.py:
```diff
@@ -88,6 +115,20 @@ def _complement_interface_empty_ip_subtrees(self):
                     iface_state[family] = {}
 
+    @property
+    def routes(self):
```
How will you call the 'running' entries if you reserve the base name to config?

cathay4t replied:
We can rename it when we need running state which I doubt it.

EdDev commented on libnmstate/state.py:
```diff
@@ -144,6 +185,24 @@ def verify_interfaces(self, other_state):
+        Verify that the self state and the other_state are identical.
+        """
+        for iface_name, routes in six.viewitems(self.iface_routes):
+            routes.sort(key=_route_sort_key)
```
This should have been done when indexing the routes (and we will not need to do this to `other` as well).

cathay4t replied:
The `metadata.py` `_generate_route_metadata()` use the `iface_routes` which does not require ordering.
The only ordering is required just for verification.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -16,12 +16,18 @@
+import socket
 
 from libnmstate import iplib
+from libnmstate.error import NmstateInternalError
```
nit: Can't you import error and use it with a context?
`error.NmstateInternalError`

cathay4t replied:
Any particular reason for this?

tyll replied:
Why is it not nice? Seems to be more like a matter of taste.

EdDev replied:
That is why it is not a must :)
In many cases, having the module as the context/namespace and avoiding for each additional error usage to add another import line is nicer IMO.

tyll replied:
After thinking some more I believe that using `from libnmstate import error` pollutes the global namespace with a generic word `error` and since it is easy to skip over many import lines I actually prefer the individual imports. Maybe moving the exceptions to libnmstate directly would make it nice enough again but I am not sure.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -16,12 +16,18 @@
+from libnmstate.error import NmstateInternalError
+from libnmstate.error import NmstateNotImplementedError
+from libnmstate.nm import nmclient
+from libnmstate.nm.nmclient import GLib
```
nit: You can just use `nmclient.GLib`

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -141,3 +147,40 @@ def _get_default_route_config(gateway, metric, default_table_id, iface_name):
+                                        IPV6_DEFAULT_GATEWAY_DESTINATION):
+            if default_gateway_defined:
+                raise NmstateNotImplementedError(
+                    'Only a single default gateway is supported.')
```
Perhaps, to avoid questions, mention that this is a per profile limitation.
I would also mention this in the commit message and in tests to show this.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -141,3 +147,40 @@ def _get_default_route_config(gateway, metric, default_table_id, iface_name):
+def _add_specfic_route(setting_ip, route):
+    destination, prefix_len = route[Route.DESTINATION].split('/')
+    prefix_len = int(prefix_len)
+    family = socket.AF_INET6 if iplib.is_ipv6_address(destination) \
```
nit: I would have prefered you use `()` instead of `\`.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -141,3 +147,40 @@ def _get_default_route_config(gateway, metric, default_table_id, iface_name):
+    ip_route = nmclient.NM.IPRoute.new(family, destination, prefix_len,
+                                       next_hop, metric)
+    if route[Route.TABLE_ID] != NM_ROUTE_TABLE_USE_DEFAULT_CFG:
+        ip_route.set_attribute(NM_ROUTE_TABLE_ATTRIBUTE,
```
Is there a BZ on this? The table should be no different from the other properties.

cathay4t replied:
The `setting_ip.props.route_table` is for gateway and DHCP routes.
The `NM_ROUTE_TABLE_ATTRIBUTE` is for specific routes.
I see no bug on this.

tyll replied:
@cathay4t would you please file a bug to request the possibility to set the default gateway the same way as other routes with the same possible settings?

cathay4t commented on libnmstate/metadata.py:
```diff
@@ -59,13 +63,16 @@ def generate_ifaces_metadata(desired_state, current_state):
         iface_state.pop(MASTER, None)
         iface_state.pop(MASTER_TYPE, None)
         iface_state.pop(BRPORT_OPTIONS, None)
+        iface_state.get(Interface.IPV4, {}).pop(ROUTES, None)
```
As I mentioned, the CI will fail on this line.

cathay4t commented:
New patch uploaded.

Changes:
 * New test case for testing multiple gateway.  That test case is marked as `pytest.xfail` due to bug https://bugzilla.redhat.com/show_bug.cgi?id=1707396
 * Some decorative changes.

EdDev commented on libnmstate/state.py:
```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
+                                route)
+
+        for route in current.routes:
+            iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
```
Why are you adding "other"/current to the "self" container?

cathay4t replied:
Merging current route into self.

cathay4t replied:
Merging other/current routes into desired routes.

EdDev replied:
But a route in the desired may have less properties than the one in current, no?
Also, it is very confusing to read... why not keep each in their own set and then perform operation on all of them... it may be even more efficient.

EdDev commented on libnmstate/state.py:
```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
+            current_iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                _Route(route))
+
+        for iface_name in list(six.viewkeys(iface_route_sets)):
```
What is the `list` all about?

cathay4t replied:
Since the `iface_route_sets` might change at `del iface_route_sets[iface_name]`, using `list()` could present the warning about change when iterating.

EdDev replied:
This is why you should not remove entries in the middle of the loop. Use `filter` or a an immutable approach to do it.

EdDev commented on libnmstate/state.py:
```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
+                del iface_route_sets[iface_name]
+
+        # Create basic interface information for changed routes.
+        for iface_name in six.viewkeys(iface_route_sets):
```
I think this method is doing more than just route state merging, it has a side affect of normalizing the interfaces state as well.
I guess you should split it and execute it explicitly in a separated method.

cathay4t replied:
New function `state.include_changed_interfaces()`

EdDev commented on libnmstate/state.py:
```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
+               iface_name in current.interfaces:
+                self.interfaces[iface_name] = {'name': iface_name}
+
+        self._state[Route.KEY] = {Route.CONFIG: []}
```
I think it is better to touch the _state only when the new data is ready.
Something like: `self._state[Route.KEY] = [route_obj.state for routes in six.viewvalue(iface_route_sets) for route_obj in routes]`

EdDev commented on libnmstate/state.py:
```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
+        the interface profile.
+        """
+        # Merge other_routes
+        iface_route_sets = defaultdict(set)
```
I am adding this comment after reading the implementation below.

Looks like you actually need indexed routes, where the entries themselves are the `_Route` object.
So, you could use that in the indexed version you currently use (convert the route entries into the object on state creation in the first place).
The `set` is also something that can be maintain from the beginning: Use a set instead of a list in the indexed structure.
Then, you just need to read it from `self` and `other`.
Note that even if you do not do that (for whatever reason), having these structures created for you by a state method fits here: `self._indexed_route_objects_by_iface().

If we separate between the self and other data/sets, then you can do whatever calculation you want in the end:
```
common = self._indexed_routes & other._indexed_routes 
all = self._indexed_routes | other._indexed_routes
...
```

cathay4t replied:
The `desired_state.iface_routes` is used by `metadata.py` which does not need to know the route
object set stuff.

Doing the converting to object sets in `__init__()` or `merge_route_config()` has no difference right now as `merge_route_config()` is the only user for those route object sets.

Changing internal information from dict to object based is another big topic which I intend to do it in another PR.

EdDev replied:
The internal representation of the data is an nmstate thing and you created an object to represent a route entry.
I do not understand your argument about why it is not used all over if it is useful.
If you have two data structures, it makes things more complex.

The claim that only the metadata is using something is not a reason to duplicate data or have two structures. If one can serve both, it makes things simpler and cleaner.

This is not a blocker, but I would prefer to see it done sooner than latter.

tyll commented on tests/integration/testlib/statelib.py:
```diff
@@ -77,7 +78,15 @@ def filter(self, based_on_state):
             if ifstate['name'] in base_iface_names
         ]
-        self._state = {INTERFACES: filtered_iface_state}
+        for rt_type in (Route.RUNNING, Route.CONFIG):
```
Please do not abbreviate variables if not necessary, `route_type` works fine here IMHO.

tyll replied:
Not sure if the name really makes sense because `type` is very generic and I would not associate this with running vs. config., I think. But I do not have a better suggestion atm.

EdDev replied:
IMO abbreviation in small scopes (few lines) is not a problem as you can see what it is immediately.
If you extract it to its our function, it would probably solve this issue and help reading in general:
```
self._state[Route.KEY][Route.RUNNING] = _filter_routes(Route.RUNNING, ...)
self._state[Route.KEY][Route.CONFIG] = _filter_routes(Route.CONFIG, ...)
```

tyll replied:
Any unnecessary abbreviation steals important brain computation time that is not available for other tasks whenever one is reading the code. The only gain is saving a few characters to type...

cathay4t replied:
I don't want to join this fight, but the linux kernel code style said:

```
LOCAL variable names should be short, and to the point.
```

That's why I am use the `rt_type` initially. Anyway, I will change to whatever way to get this drag-along PR merged.

tyll commented on libnmstate/nm/route.py:
```diff
@@ -141,3 +146,45 @@ def _get_default_route_config(gateway, metric, default_table_id, iface_name):
+                                        IPV6_DEFAULT_GATEWAY_DESTINATION):
+            if default_gateway_defined:
+                raise NmstateNotImplementedError(
+                    'Only a single default gateway is supported due to '
```
This is handling a single profile, isn't it? Not sure what you are getting at...

EdDev replied:
So you are saying that this information should be decoded from the trace and not from the message?
I think it is confusing, but I leave it for the composer discretion, the commit message is good enough.

tyll replied:
> So you are saying that this information should be decoded from the trace and not from the message?
> I think it is confusing, but I leave it for the composer discretion, the commit message is good enough.

I am still not sure what you are suggesting... Are you proposing to update the message to mention `Only a single default gateway per interface and IP family is supported...`? It could also mention the Bugzilla entry to be filed to allow to specify gateways like normal routes.

EdDev replied:
I am proposing something like: `Only a single default gateway is supported per interface due to...`
BZ/s are mentioned in the commit message, logs do not include such info.

tyll commented on libnmstate/state.py:
```diff
@@ -30,6 +32,31 @@
+
+class _Route(object):
+    def __init__(self, route):
+        self.state = {
```
I assumed this is the sub-state for just a route entry, so it will need to be created anyhow.

EdDev replied:
Thanks for the clarification.
Processing data using attributes seems to be more appropriate to me. The way the data is represented is a different manner.
Note that using dicts will also cause the data to be shared by two sources (one dict, multiple pointers to it). I do not know the implications of that in this case. I just found it surprising and that is why I asked.

EdDev replied:
@tyll, you are right about the reference, no idea what I was thinking at the time.
I guess it ends up to the question of how data should be preserved in the object.
I specifically do not think this is a nice state representation, instead of referencing the specific attribute, there is a need to look it up in the dict. The hashing as an example, needs to dig in the dict instead of listing the attributes...
See https://github.com/nmstate/nmstate/pull/345/files#diff-cf08a4a46976f25e68653866cd783ebbR26

tyll commented on libnmstate/nm/route.py:
```diff
@@ -16,12 +16,17 @@
+from libnmstate.schema import Route
 
-NM_MAIN_ROUTE_TABLE_ID = 0
+NM_ROUTE_TABLE_USE_DEFAULT_CFG = 0
```
IMHO it should be `CONFIG` instead of `CFG`.

tyll replied:
Alternative name: `NM_USE_DEFAULT_ROUTE_TABLE`.

tyll replied:
Or `NM_ROUTE_USE_DEFAULT_TABLE`.

tyll commented on tests/integration/route_test.py:
```diff
@@ -0,0 +1,208 @@
+def _assert_has_route(route, state):
+    assert route in state['routes']['config']
+
+    # For running route, the default(-1) metric and route-table will be
```
There needs to be a space character before the parenthesis.

tyll commented on tests/integration/route_test.py:
```diff
@@ -0,0 +1,208 @@
+
+def _assert_not_has_route(route, state):
+    for routes in (state['routes']['running'], state['routes']['config']):
+        assert route not in routes
```
Why does this not need to handle metric and table-id in a special way as the `_assert_has_route()` function does?

cathay4t replied:
Yeah. I should do the same. But I have removed `_assert_has_route()` from this PR as it's not used.

tyll commented on libnmstate/metadata.py:
```diff
@@ -152,3 +159,23 @@ def _generate_link_master_metadata(ifaces_desired_state,
+        iface_state = desired_state.interfaces.get(iface_name, {})
+        for family in (Interface.IPV4, Interface.IPV6):
+            if family in iface_state:
+                iface_state[family]['_routes'] = []
```
`_routes` should probably be a constant, also in the other lines.

tyll commented on libnmstate/metadata.py:
```diff
@@ -152,3 +159,23 @@ def _generate_link_master_metadata(ifaces_desired_state,
+            else:
+                iface_state[family] = {'_routes': []}
+        for route in routes:
+            if iplib.is_ipv6_address(route[Route.DESTINATION]):
```
Suggestion:

```suggestion
            if iplib.is_ipv6_address(route[Route.DESTINATION]):
                family = Interface.IPV6
            else:
                family = Interface.IPV4
            iface_state[family].setdefault('_routes', []).append(route)
```

Then you can remove the code above to initialize the routes metadata with an empty list. Please note, that `_routes` here should be replace with the constant that should be introduced IMHO.

cathay4t replied:
Still need to handle `family not in iface_state` and calling `setdefault()` for every route seems time wasting.

tyll replied:
AFAIU #340 `family in iface_state` is always true now. Does `setdefault()` have a serious performance impact over `[ROUTES]`? The latter is also calling `__getitem__()`.

EdDev commented on tests/integration/route_test.py:
```diff
@@ -0,0 +1,208 @@
+    # changed. Hence should be ignored.
+    tmp_route = copy.deepcopy(route)
+    run_routes = copy.deepcopy(state['routes']['running'])
+    for prop_name, default_value in (('metric', -1), ('table-id', 0)):
```
I do not really understand where this scenario exists in the tests. The defaults seem to be checked against the desired state.
I do not expect the client to input in the desired state such values, as they mean nothing. If the client does not specify them, nmstate may internally use some value to represent it, but that is an implementation detail.
Is there something I am missing here?

Maybe you mean what is reported back in the 'config' state? I expect it will return either nothing or the actual explicit value. We cannot expect the client to have this logic.

cathay4t replied:
When desire state does not define metric or table_id. The config will be -1 and 0, the running will be unknown value. To do asserting, we need to assume above is legal.

EdDev commented on tests/lib/netapplier_test.py:
```diff
@@ -24,6 +24,7 @@
 from libnmstate.schema import Constants
 
 INTERFACES = Constants.INTERFACES
+ROUTES = Constants.ROUTES
```
Leftover?

tyll commented on tests/integration/testlib/statelib.py:
```diff
@@ -77,7 +78,15 @@ def filter(self, based_on_state):
+            ]
+            if filtered_routes:
+                self._state[Route.KEY][rt_type] = filtered_routes
+        self._state[INTERFACES] = filtered_iface_state
```
It is confusing that this line comes after the route filtering. IMHO it should stay before the route filtering. Also the change from not using `INTERFACES: ...` to the new form would have been a perfect change for a commit with the related changes to use `netinfo.show()`.

cathay4t commented on libnmstate/state.py:
```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
+                iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                    _Route(route))
+            else:
+                logging.warning("Ignoring the route entry with no "
```
It got discard which will not fail the verification.

EdDev replied:
Discard a route? That is not handled or tested in this PR.

cathay4t commented on libnmstate/state.py:
```diff
@@ -264,3 +374,9 @@ def dict_update(origin_data, to_merge_data):
+
+
+def _route_sort_key(route):
+    return (route.get(Route.TABLE_ID, -1),
```
Changed to `nm_route.NM_ROUTE_TABLE_USE_DEFAULT_CFG`

cathay4t commented:
> I somehow missed that the state has no unit tests.

Why I need to do unit test when integration test cases covers all the functions of them?

cathay4t commented on tests/integration/route_test.py:
```diff
@@ -0,0 +1,208 @@
+                ]
+            },
+        })
+    cur_state = statelib.show_only(('eth1',))
```
Cause above routes is for `eth1` only.

EdDev replied:
But you are not checking the `eth` interface, but the routes.

cathay4t replied:
Just want to filter out the routes from other interfaces.

cathay4t commented:
New patch uploaded.
Changes:
 * Add new test cases: `test_add_route_without_metric` and `test_add_route_without_table_id`.
 * Trivial decorative changes.

EdDev commented:
> Why I need to do unit test when integration test cases covers all the functions of them?

There are many types of tests, each has their usefulness and limitations.
Unit tests is an integral part of development, keeping a very close loop with the changing code. When the code is written for the first time and when the code changes. The cases where unit tests are skipped are limited to when they require too many mocks (like the nm sub-package) with no other alternatives. The exception is not having unit tests, not the other way around.
Yes, there are many arguments against unit tests, TDD and others, you can open this out in an email thread if you want.

I would recommend writing unit tests as part of your coding progress (pre or post, whatever works out for you), but if you need help with this specific one, please let me know and I'll write some for the status module.

cathay4t commented:
New patch uploaded.

Changes:
 * Add new test case: `test_add_route_without_iface_state`. It just test adding routes to interface is not defined in `state['interfaces']`.

cathay4t commented:
> > Why I need to do unit test when integration test cases covers all the functions of them?
> 
> There are many types of tests, each has their usefulness and limitations.
> Unit tests is an integral part of development, keeping a very close loop with the changing code.

For this PR, Could you name a unit test case I need to write?
When I try to add unit test case for `state_test.py`, I found myself duplicating what I did in integration test.


> Yes, there are many arguments against unit tests, TDD and others, you can open this out in an email thread if you want.
No. I will not discuss topic like vim vs emac.

cathay4t commented:
New patch uploaded:

Changes:
 * New unit test cases added:
    * `TestAssertRouteState::test_merge_route_add`
    * `TestAssertRouteState::test_merge_route_add`
    * `TestAssertRouteState::test_merge_route_add_without_iface`
    * `TestAssertRouteState::test_verify_routes_add_route`
    * `TestAssertRouteState::test_add_route_without_metric`
    * `TestAssertRouteState::test_add_route_without_route_table_id`
 * New constants:
    * `libnmstate.schema.Route.USE_DEFAULT_METRIC`
    * `USE_DEFAULT_ROUTE_TABLE`

EdDev commented on tests/lib/state_test.py:
```diff
@@ -138,3 +140,122 @@ def _extra_state(self):
+        assert (new_route[Route.NEXT_HOP_INTERFACE] in
+                desired_state.interfaces)
+
+    def test_verify_routes_add_route(self):
```
How about: `test_routes_verification_ignores_order`.

In the test body, please reverse the list instead of using a tmp.
You can use an immutable ([::-1]) or mutable (list.reverse()) version.

EdDev commented on tests/lib/state_test.py:
```diff
@@ -138,3 +140,122 @@ def _extra_state(self):
+        desired_state = state.State(desired_state_dict)
+        desired_state.verify_routes(current_state)
+
+    def test_add_route_without_metric(self):
```
I do not understand what happens here.
To create the new state, a route is copied, its metric field is removed, interfaces subtree removed (why?), then after the merge, a default metric is added to the new route (in practice overwriting what was there before, embedded in the new state object). Finally the new route is checked to exist in the destination.

You better check what exactly exists in the new state after the merge, perhaps something got overwritten/removed.
There is also a matter of checking against the general state (s.state) and not only the indexed output (relevant to all tests).

cathay4t commented:
I am tried of these. Will open new PR when I have the mod.
